### PR TITLE
Enable binary build - migrate crypto library APIs + refactor

### PR DESCRIPTION
**Pull request by hMcLauchlan**
This PR has two changes. A very small test fix and a very large refactor / rewrite.

Please see commit message for details.

I have a few lingering doubts about this PR, which I'll point out explicitly. I'd love to hear your thoughts:
- I'm not too sure about the `Sha...` -> `Sha2_...` renaming. I personally feel like it improves clarity (and brings sha2 in line with sha3 naming convention, while distinguishing from Sha1, which AFAICT isn't in this codebase), but it did add a lot of LOC to this patch that might not have needed to be there.
- While rewriting the `Sha2` and `Sha3` files, I omitted a lot of comments that were present in the original. I felt that by leaving detailed comments in the header and having comments for major, meaningful functions, there was no need for comprehensive comments on thin wrapper functions. I.e. I commented `HashAllGeneric` but not its wrappers. I'm open to adding them if you feel like it's a good idea.
- I've tried to follow the C convention in this codebase as best as I can, but it's a flavor of C I'm not really used to. I apologize for the weird style mistakes; I couldn't find any style guide/linters handy in the repo.

Next step is probably to finish up cleaning `Hash/` by porting the SM3 stuff. Then I'll likely do a patch just to root out internal headers. After which I should be able to start making mostly Makefile/GnuMake/Cmake changes.

Looking forward to the review, thanks!

### CODE REVIEW

File: OsStub/BaseCryptLibMbedTls/Hash/CryptSha2.c [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+  Computes the size of a mbedtls_sha{256, 512}_context struct given a
+  Sha2Type. Note that 384 and 512 both use mbedtls_sha512_context.
+
+  @param[in]   type     Algorithm requested. Must be supported.
+
+  @retval UINTN > 0   The size of the corresponding context struct.
+  @retval UINTN == 0   Error; algorithm not supported.
+
+**/
+static UINTN GetContextSize(enum InternalSha2Type type)
```
Comment (jyao1) [inline OsStub/BaseCryptLibMbedTls/Hash/CryptSha2.c]: This coding style is not acceptable.

Please use:

STATIC
UINTN
GetContextSize (
  IN INTERNAL_SHA2_TYPE Type
  )

File: OsStub/BaseCryptLibMbedTls/Hash/CryptSha2.c [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+Copyright (c) 2009 - 2016, Intel Corporation. All rights reserved.<BR>
+SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include "InternalCryptLib.h"
+#include <mbedtls/sha256.h>
+#include <mbedtls/sha512.h>
+
+enum InternalSha2Type{_SHA2_256, _SHA2_384, _SHA2_512};
```
Comment (jyao1) [inline OsStub/BaseCryptLibMbedTls/Hash/CryptSha2.c]: This coding style is not acceptable.

Please use:
typedef enum {
  _SHA2_256,
  _SHA2_384,
  _SHA2_512,
} INTERNAL_SHA2_TYPE;

File: OsStub/BaseCryptLibMbedTls/Hash/CryptSha2.c [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+  @param[in]   type     Algorithm requested. Must be supported.
+
+  @retval UINTN > 0   The size of the corresponding context struct.
+  @retval UINTN == 0   Error; algorithm not supported.
+
+**/
+static UINTN GetContextSize(enum InternalSha2Type type)
+{
+  switch (type)
+  {
```
Comment (jyao1) [inline OsStub/BaseCryptLibMbedTls/Hash/CryptSha2.c]: please use switch (type) {

File: OsStub/BaseCryptLibMbedTls/Hash/CryptSha2.c [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+
+**/
+static VOID* InitAndStartShaContext(enum InternalSha2Type type)
+{
+  VOID* ctx;
+
+  ctx = (VOID*)AllocatePool(GetContextSize(type));
+
+  if (!ctx)
+    return FALSE;
```
Comment (jyao1) [inline OsStub/BaseCryptLibMbedTls/Hash/CryptSha2.c:77]: This coding style is not acceptable.

Please use :
if (!ctx) {
  return FALSE;
}

File: OsStub/BaseCryptLibMbedTls/Hash/CryptSha2.c [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+
+  return TRUE;
+}
+
+BOOLEAN
+EFIAPI
+Sha2_256Init
+(
+  OUT VOID **ShaCtxPtr
+)
```
Comment (jyao1) [inline OsStub/BaseCryptLibMbedTls/Hash/CryptSha2.c:189]: please use:

Sha2_256Init (
  OUT VOID **ShaCtxPtr
  )

File: OsStub/BaseCryptLibOpenssl/Hash/CryptSm3.c [hunk 6]
```diff
@@ -209,26 +209,28 @@ Sm3HashAll (
...
 {
-  SM3_CTX Ctx;
+  INTN success;
+  EVP_MD_CTX* ctx;
 
   //
   // Check input parameters.
   //
-  if (HashValue == NULL) {
+  if (!HashValue)
```
Comment (jyao1) [inline OsStub/BaseCryptLibOpenssl/Hash/CryptSm3.c]: This is not acceptable.
HashValue is *UINT8, not boolean.

Please change it back.

File: OsStub/BaseCryptLibOpenssl/Hash/CryptSm3.c [hunk 7]
```diff
@@ -209,26 +209,28 @@ Sm3HashAll (
...
-  }
 
-  //
-  // SM3 Hash Computation.
-  //
-  sm3_init(&Ctx);
+  ctx = EVP_MD_CTX_new();
 
-  sm3_update(&Ctx, Data, DataSize);
+  if (!ctx)
```
Comment (jyao1) [inline OsStub/BaseCryptLibOpenssl/Hash/CryptSm3.c]: Ditto, please if (ctx == NULL);

File: OsStub/BaseCryptLibOpenssl/Hash/CryptSm3.c [hunk 8]
```diff
@@ -209,26 +209,28 @@ Sm3HashAll (
...
+    return FALSE;
+
+  success = EVP_Digest(Data, DataSize, HashValue, NULL /*bytes written*/,
+                      EVP_sm3(), NULL /*impl*/);
 
-  sm3_final(HashValue, &Ctx);
+  if (!success)
```
Comment (jyao1) [inline OsStub/BaseCryptLibOpenssl/Hash/CryptSm3.c]: Ditto, please use if (success == 0)

**Comment by jyao1** [comment]
Thanks for the patch. I see two problem here.

1) (It is mentioned in readme.md) We are moving and reorganizing the openspdm repo and we freeze the master.
I created https://github.com/jyao1/openspdm/tree/temp_master to achieve all activities during the code freeze. And I will resync all patches from temp_master to master after that is done.

So, please create a pull-request to temp_master, instead of master.

2) Coding style is an big issue. In openspdm, we are following the EDKII coding style. The naming is following camel case, we are changing to snake case after reorg.

I advise some general coding style issues:
2.1) if (XXX) v.s. if (XXX == AAA).
If (XXX) is only used when XXX is bool.
If XXX is other type, please use (XXX == AAA)

2.2) if (XXX) Statement; v.s if (XXX) {Statement;}
"if (XXX) Statement" is forbidden in any case, event Statement is just one line.
Please always use "if (XXX) {Statement}" - with brace around.

Please try to look at current code and use same style when you write new code.

**Comment by jyao1** [comment]
Is this a user mistake?
I recommend to have PR for temp_master, not master.

**Comment by hMcLauchlan** [comment]
I think github automatically updated this pr when I updated the branch. I was in the middle of opening a new PR. Closing now.

**Pull request closed without merging**